crates/in_stream/src/ws.rs

```diff
@@ -0,0 +1,412 @@
+use crate::*;
+use std::io::{Error, ErrorKind, Result};
+use url2::prelude::*;
+
+mod frame;
+pub use frame::*;
+
+const SCHEME: &'static str = "wss";
+
+/// internal helper, make sure we're dealing with wss urls
+fn validate_url_scheme(url: &Url2) -> Result<()> {
+    if url.scheme() != SCHEME {
+        return Err(Error::new(
+            ErrorKind::InvalidInput,
+            format!("got: '{}', expected: '{}://...'", SCHEME, url),
+        ));
+    }
+    Ok(())
+}
+
+#[derive(Debug)]
+/// websocket specific bind configuration
+pub struct WssBindConfig {
+    pub sub_bind_config: InStreamConfigAny,
+}
+
+impl WssBindConfig {
+    pub fn new<Sub: InStreamConfig>(sub_config: Sub) -> Self {
+        Self {
+            sub_bind_config: sub_config.to_any(),
+        }
+    }
+}
+
+impl InStreamConfig for WssBindConfig {}
+
+/// bind to a network interface to listen for websocket connections
+#[derive(Debug)]
+pub struct InStreamListenerWss<Sub: InStreamListenerStd> {
+    sub: Sub,
+}
+
+impl<Sub: InStreamListenerStd> InStreamListenerWss<Sub> {
+    pub fn bind(url: &Url2, config: WssBindConfig) -> Result<Self> {
+        InStreamListenerWss::raw_bind(url, config)
+    }
+}
+
+impl<Sub: InStreamListenerStd> InStreamListener<&mut WsFrame, WsFrame>
+    for InStreamListenerWss<Sub>
+{
+    type Stream = InStreamWss<Sub::StreamStd>;
+
+    fn raw_bind<C: InStreamConfig>(url: &Url2, config: C) -> Result<Self> {
+        let config = WssBindConfig::from_gen(config)?;
+        validate_url_scheme(url)?;
+        let mut url = url.clone();
+        url.set_scheme(Sub::StreamStd::URL_SCHEME).unwrap();
```

Comment:
Why not use `?` here instead?

Reply:
can't fail unless scheme is mal-formed - and scheme is a constant... will add a comment.